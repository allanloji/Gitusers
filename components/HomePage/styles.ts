import styled from '@emotion/styled';

const ContentContainer = styled.div`
  position: relative;
  padding: 4rem 6rem;
  max-width: 1500px;
  margin: 0 auto;
`;

const UsersContainer = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  width: 100%;
  gap: 2rem 2.5rem;

  @media (max-width: 992px) {
    grid-template-columns: 1fr 1fr;
  }

  @media (max-width: 768px) {
    grid-template-columns: 1fr;
  }
`;

export { ContentContainer, UsersContainer };
