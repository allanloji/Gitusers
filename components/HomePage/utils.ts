const perPageOptions = [
  { label: '9 per page', value: '9' },
  { label: '18 per page', value: '18' },
  { label: '27 per page', value: '27' },
];

const sortOptions = [
  { label: 'Dont sort', value: '' },
  { label: 'By # of followers', value: 'followers' },
  { label: 'By # of repositories', value: 'repositories' },
  { label: 'By date joined', value: 'joined' },
];

const orderOptions = [
  { label: 'Any order', value: '' },
  { label: 'Asc', value: 'asc' },
  { label: 'Desc', value: 'desc' },
];

export { perPageOptions, sortOptions, orderOptions };
