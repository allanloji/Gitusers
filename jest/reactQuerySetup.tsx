import React from 'react';
import { QueryClientProvider, setLogger, QueryClient } from 'react-query';

const setupReactQuery = () => {
  setLogger({
    // eslint-disable-next-line no-console
    log: console.log,
    // eslint-disable-next-line no-console
    warn: console.warn,
    error: () => {},
  });

  const queryClient = new QueryClient({
    defaultOptions: {
      queries: { retry: false, staleTime: 0 },
    },
  });

  // @ts-ignore
  const wrapper = ({ children }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  return { wrapper, queryClient };
};

export default setupReactQuery;
