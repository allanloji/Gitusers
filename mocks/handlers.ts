import searchHandlers from './search';

export const handlers = [...searchHandlers];
