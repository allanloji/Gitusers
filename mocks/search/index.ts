import usersHandler from './users';

const handlers = [usersHandler];

export default handlers;
