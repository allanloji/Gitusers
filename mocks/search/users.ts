// eslint-disable-next-line import/no-extraneous-dependencies
import { rest } from 'msw';

const response = {
  total_count: 84430,
  incomplete_results: false,
  items: [
    {
      login: 'test123',
      id: 27031,
      node_id: 'MDQ6VXNlcjI3MDMx',
      avatar_url: 'https://avatars.githubusercontent.com/u/27031?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/test123',
      html_url: 'https://github.com/test123',
      followers_url: 'https://api.github.com/users/test123/followers',
      following_url: 'https://api.github.com/users/test123/following{/other_user}',
      gists_url: 'https://api.github.com/users/test123/gists{/gist_id}',
      starred_url: 'https://api.github.com/users/test123/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/test123/subscriptions',
      organizations_url: 'https://api.github.com/users/test123/orgs',
      repos_url: 'https://api.github.com/users/test123/repos',
      events_url: 'https://api.github.com/users/test123/events{/privacy}',
      received_events_url: 'https://api.github.com/users/test123/received_events',
      type: 'User',
      site_admin: false,
      score: 1.0,
    },
    {
      login: 'testjie',
      id: 23480729,
      node_id: 'MDQ6VXNlcjIzNDgwNzI5',
      avatar_url: 'https://avatars.githubusercontent.com/u/23480729?v=4',
      gravatar_id: '',
      url: 'https://api.github.com/users/testjie',
      html_url: 'https://github.com/testjie',
      followers_url: 'https://api.github.com/users/testjie/followers',
      following_url: 'https://api.github.com/users/testjie/following{/other_user}',
      gists_url: 'https://api.github.com/users/testjie/gists{/gist_id}',
      starred_url: 'https://api.github.com/users/testjie/starred{/owner}{/repo}',
      subscriptions_url: 'https://api.github.com/users/testjie/subscriptions',
      organizations_url: 'https://api.github.com/users/testjie/orgs',
      repos_url: 'https://api.github.com/users/testjie/repos',
      events_url: 'https://api.github.com/users/testjie/events{/privacy}',
      received_events_url: 'https://api.github.com/users/testjie/received_events',
      type: 'User',
      site_admin: false,
      score: 1.0,
    },
  ],
};

const handler = rest.get('https://api.github.com/search/users', (req, res, ctx) =>
  res(
    // Respond with a 200 status code
    ctx.status(200),
    ctx.json(response),
  ),
);

export { response };
export default handler;
