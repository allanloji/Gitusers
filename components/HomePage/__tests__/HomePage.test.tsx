import React from 'react';
import { queryAllByRole, render, waitForElementToBeRemoved } from '@testing-library/react';
import { rest } from 'msw';

import reactQuerySetup from 'jest/reactQuerySetup';
import { server } from 'mocks/server';
import HomePage from 'pages';
import userEvent from '@testing-library/user-event';
import { emptyResponse } from 'mocks/search/users';

const { wrapper, queryClient } = reactQuerySetup();

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  queryClient.clear();
});
afterAll(() => server.close());

describe('HomePage ', () => {
  test('Should render', () => {
    // @ts-ignore
    const { container } = render(<HomePage />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  test('Should render users', async () => {
    const { container, getByPlaceholderText, getByTestId } = render(<HomePage />, {
      // @ts-ignore
      wrapper,
    });

    const input = getByPlaceholderText('Search by name');
    const submit = getByTestId('submit-bttn');

    userEvent.type(input, 'Test');
    userEvent.click(submit);

    const skeleton = getByTestId('skeleton');

    await waitForElementToBeRemoved(skeleton);

    const selects = queryAllByRole(container, 'combobox');

    userEvent.selectOptions(selects[0], '9 per page');
    userEvent.selectOptions(selects[1], 'By date joined');
    userEvent.selectOptions(selects[2], 'Asc');

    expect(container).toMatchSnapshot();
  });

  test('Should render message when theres no users', async () => {
    server.use(
      rest.get('https://api.github.com/search/users', (req, res, ctx) =>
        res(ctx.json(emptyResponse)),
      ),
    );

    const { container, getByPlaceholderText, getByTestId } = render(<HomePage />, {
      // @ts-ignore
      wrapper,
    });

    const input = getByPlaceholderText('Search by name');
    const submit = getByTestId('submit-bttn');

    userEvent.type(input, 'Test');
    userEvent.click(submit);

    const skeleton = getByTestId('skeleton');

    await waitForElementToBeRemoved(skeleton);

    expect(container).toMatchSnapshot();
  });

  test('Should render error message', async () => {
    server.use(
      rest.get('https://api.github.com/search/users', (req, res, ctx) => res(ctx.status(401))),
    );

    const { container, getByPlaceholderText, getByTestId } = render(<HomePage />, {
      // @ts-ignore
      wrapper,
    });

    const input = getByPlaceholderText('Search by name');
    const submit = getByTestId('submit-bttn');

    userEvent.type(input, 'Test');
    userEvent.click(submit);

    const skeleton = getByTestId('skeleton');

    await waitForElementToBeRemoved(skeleton);

    expect(container).toMatchSnapshot();
  });
});
