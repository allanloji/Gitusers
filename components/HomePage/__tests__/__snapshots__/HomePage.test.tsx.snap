// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage  Should render 1`] = `
<div>
  <div
    class="css-1op182y"
  >
    <h1
      class="css-1i1wdm8"
    >
      gitusers 
    </h1>
    <span
      class="css-dkt9vh"
    >
      👦🏻
    </span>
    <h2
      class="css-vkg5qd"
    >
      A github user search
    </h2>
  </div>
  <div
    class="css-16fhlax"
  >
    <div
      class="css-1wisk8h"
    >
      <input
        class="css-1bfqk3l"
        placeholder="Search by name"
        value=""
      />
      <button
        class="css-zq0wcq"
        data-testid="submit-bttn"
      >
        <img
          alt="Search icon"
          class="css-ulohxf"
          src="/assets/search.svg"
        />
      </button>
    </div>
    <div
      class="css-1dszs7t"
    >
      <span
        class="css-gjmu0c"
      >
        💡
      </span>
      <h1
        class="css-17s2ops"
      >
        Search a user in order to see results
      </h1>
    </div>
  </div>
</div>
`;

exports[`HomePage  Should render error message 1`] = `
<div>
  <div
    class="css-1op182y"
  >
    <h1
      class="css-1i1wdm8"
    >
      gitusers 
    </h1>
    <span
      class="css-dkt9vh"
    >
      👦🏻
    </span>
    <h2
      class="css-vkg5qd"
    >
      A github user search
    </h2>
  </div>
  <div
    class="css-16fhlax"
  >
    <div
      class="css-1wisk8h"
    >
      <input
        class="css-1bfqk3l"
        placeholder="Search by name"
        value="Test"
      />
      <button
        class="css-zq0wcq"
        data-testid="submit-bttn"
      >
        <img
          alt="Search icon"
          class="css-ulohxf"
          src="/assets/search.svg"
        />
      </button>
    </div>
    <div
      class="css-1dszs7t"
    >
      <span
        class="css-gjmu0c"
      >
        🤖
      </span>
      <h1
        class="css-17s2ops"
      >
        There was an error getting the users or API limit
      </h1>
    </div>
  </div>
</div>
`;

exports[`HomePage  Should render message when theres no users 1`] = `
<div>
  <div
    class="css-1op182y"
  >
    <h1
      class="css-1i1wdm8"
    >
      gitusers 
    </h1>
    <span
      class="css-dkt9vh"
    >
      👦🏻
    </span>
    <h2
      class="css-vkg5qd"
    >
      A github user search
    </h2>
  </div>
  <div
    class="css-16fhlax"
  >
    <div
      class="css-1wisk8h"
    >
      <input
        class="css-1bfqk3l"
        placeholder="Search by name"
        value="Test"
      />
      <button
        class="css-zq0wcq"
        data-testid="submit-bttn"
      >
        <img
          alt="Search icon"
          class="css-ulohxf"
          src="/assets/search.svg"
        />
      </button>
    </div>
    <div
      class="css-1dszs7t"
    >
      <span
        class="css-gjmu0c"
      >
        👀
      </span>
      <h1
        class="css-17s2ops"
      >
        There are no users with that name
      </h1>
    </div>
  </div>
</div>
`;

exports[`HomePage  Should render users 1`] = `
<div>
  <div
    class="css-1op182y"
  >
    <h1
      class="css-1i1wdm8"
    >
      gitusers 
    </h1>
    <span
      class="css-dkt9vh"
    >
      👦🏻
    </span>
    <h2
      class="css-vkg5qd"
    >
      A github user search
    </h2>
  </div>
  <div
    class="css-16fhlax"
  >
    <div
      class="css-1wisk8h"
    >
      <input
        class="css-1bfqk3l"
        placeholder="Search by name"
        value="Test"
      />
      <button
        class="css-zq0wcq"
        data-testid="submit-bttn"
      >
        <img
          alt="Search icon"
          class="css-ulohxf"
          src="/assets/search.svg"
        />
      </button>
    </div>
    <p
      class="css-1l7d2vk"
    >
      About 
      84430
       results (first 1000 displayed due to API limit)
    </p>
    <div
      class="css-12xxxyo"
    >
      <select
        class="css-1fhsdaz"
        id="perPage"
        name="perPage"
      >
        <option
          value="9"
        >
          9 per page
        </option>
        <option
          value="18"
        >
          18 per page
        </option>
        <option
          value="27"
        >
          27 per page
        </option>
      </select>
      <select
        class="css-1fhsdaz"
        id="sort"
        name="sort"
      >
        <option
          value=""
        >
          Dont sort
        </option>
        <option
          value="followers"
        >
          By # of followers
        </option>
        <option
          value="repositories"
        >
          By # of repositories
        </option>
        <option
          value="joined"
        >
          By date joined
        </option>
      </select>
      <select
        class="css-1fhsdaz"
        id="order"
        name="order"
      >
        <option
          value=""
        >
          Any order
        </option>
        <option
          value="asc"
        >
          Asc
        </option>
        <option
          value="desc"
        >
          Desc
        </option>
      </select>
    </div>
    <div
      class="css-i7kby6"
    >
      <a
        href="https://github.com/test123"
        rel="noreferrer noopener"
        target="_blank"
      >
        <div
          class="css-1gi9ab3"
        >
          <div
            class="css-1jma1eq"
            color="#90BE6D"
          >
            <img
              alt="Avatar url for test123"
              class="css-a6g4fg"
              src="https://avatars.githubusercontent.com/u/27031?v=4"
            />
          </div>
          <div
            class="css-1obzys2"
          >
            <p
              class="css-1mceto1"
            >
              test123
            </p>
            <p
              class="css-19cwc4a"
            >
              User
            </p>
          </div>
        </div>
      </a>
      <a
        href="https://github.com/testjie"
        rel="noreferrer noopener"
        target="_blank"
      >
        <div
          class="css-1gi9ab3"
        >
          <div
            class="css-1p5pmkk"
            color="#277DA1"
          >
            <img
              alt="Avatar url for testjie"
              class="css-a6g4fg"
              src="https://avatars.githubusercontent.com/u/23480729?v=4"
            />
          </div>
          <div
            class="css-1obzys2"
          >
            <p
              class="css-1mceto1"
            >
              testjie
            </p>
            <p
              class="css-19cwc4a"
            >
              User
            </p>
          </div>
        </div>
      </a>
    </div>
    <div
      class="css-1mkvne5"
    >
      <button
        class="css-q1rq5h"
        data-testid="firstPage"
        disabled=""
        type="button"
      >
        &lt;&lt;
      </button>
      <button
        class="css-q1rq5h"
        data-testid="prevPage"
        disabled=""
        type="button"
      >
        &lt;
      </button>
      <p
        class="css-v8p8i1"
      >
        1
         of 
        112
      </p>
      <button
        class="css-q1rq5h"
        data-testid="nextPage"
        disabled=""
        type="button"
      >
        &gt;
      </button>
      <button
        class="css-q1rq5h"
        data-testid="lastPage"
        disabled=""
        type="button"
      >
        &gt;&gt;
      </button>
    </div>
  </div>
</div>
`;
